packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -48,49 +49,80 @@ class MouseTrackerAnnotation {
+    return '${describeIdentity(this)}(callbacks: $describeCallbacks)';
   }
 }
 
-// Used internally by the MouseTracker for accounting for which annotation is
-// active on which devices inside of the MouseTracker.
-class _TrackedAnnotation {
-  _TrackedAnnotation(this.annotation);
-
-  final MouseTrackerAnnotation annotation;
-
-  /// Tracks devices that are currently active for this annotation.
-  ///
-  /// If the mouse pointer corresponding to the integer device ID is
-  /// present in the Set, then it is currently inside of the annotated layer.
-  ///
-  /// This is used to detect layers that used to have the mouse pointer inside
-  /// them, but now no longer do (to facilitate exit notification).
-  Set<int> activeDevices = <int>{};
-}
-
-/// Describes a function that finds an annotation given an offset in logical
-/// coordinates.
+/// Signature for searching for [MouseTrackerAnnotation]s at the given offset.
 ///
 /// It is used by the [MouseTracker] to fetch annotations for the mouse
 /// position.
 typedef MouseDetectorAnnotationFinder = Iterable<MouseTrackerAnnotation> Function(Offset offset);
 
-/// Keeps state about which objects are interested in tracking mouse positions
-/// and notifies them when a mouse pointer enters, moves, or leaves an annotated
-/// region that they are interested in.
+// Various states of each connected mouse device.
+//
+// It is used by [MouseTracker] to compute which callbacks should be triggered
+// by each event.
+class _MouseState {
+  _MouseState({
+    @required PointerEvent mostRecentEvent,
+  }) : assert(mostRecentEvent != null),
+       _mostRecentEvent = mostRecentEvent;
+
+  // The list of annotations that contains this device during the last frame.
+  LinkedHashSet<MouseTrackerAnnotation> lastAnnotations =
+    LinkedHashSet<MouseTrackerAnnotation>();
+
+  // The most recent mouse event observed from this device.
+  //
+  // The [mostRecentEvent] is never null.
+  PointerEvent get mostRecentEvent => _mostRecentEvent;
+  PointerEvent _mostRecentEvent;
+  set mostRecentEvent(PointerEvent value) {
+    assert(value != null);
+    assert(value.device == _mostRecentEvent.device);
+    _mostRecentEvent = value;
+  }
+
+  int get device => _mostRecentEvent.device;
+
+  @override
+  String toString() {
+    final String describeEvent = '${_mostRecentEvent.runtimeType}(device: ${_mostRecentEvent.device})';
+    final String describeAnnotations = '[list of ${lastAnnotations.length}]';
+    return '${describeIdentity(this)}(event: $describeEvent, annotations: $describeAnnotations)';
+  }
+}
+
+/// Maintains the relationship between mouse devices and
+/// [MouseTrackerAnnotation]s, and notifies interested callbacks of the changes
+/// thereof.
 ///
 /// This class is a [ChangeNotifier] that notifies its listeners if the value of
 /// [mouseIsConnected] changes.
 ///
-/// Owned by the [RendererBinding] class.
+/// This class is a global singleton and owned by the [RendererBinding] class.
```

Comment:
nit: The class itself is not a global singelton, I can easily instantiate more than one instance. Maybe rephrase this to say that RendererBinding owns a global singelton of this class?

Reply:
I changed it to:
```dart
/// An instance of [MouseTracker] is owned by the global singleton of
/// [RendererBinding].
```